src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryScheduler.java

```diff
@@ -71,85 +69,36 @@ public RepairScheduler(
     public synchronized List<OfferID> resourceOffers(SchedulerDriver driver, List<Offer> offers, Block block)
             throws Exception {
         List<OfferID> acceptedOffers = new ArrayList<>();
-        List<TaskInfo> terminatedTasks = new ArrayList<>(getTerminatedTasks(block));
+        updateRecoveryPools(block);
 
-        // Compute the stopped and failed pool for UI
-        RepairStatus oldStatus = repairStatusRef.get();
-        if (oldStatus == null) {
-            oldStatus = new RepairStatus(Collections.emptyList(), Collections.emptyList());
-        }
-        List<String> oldFailed = oldStatus.getFailed();
-        List<String> newFailed = new ArrayList<>(terminatedTasks.stream()
-                .filter(failureMonitor::hasFailed)
-                .map(TaskInfo::getName)
-                .collect(Collectors.toList()));
-        newFailed.addAll(oldFailed);
-        //TODO (ghartman) possibly write this list down in ZK, to ensure the UI doesn't change upon restart
-        newFailed = newFailed.stream().distinct().collect(Collectors.toList());
-        List<String> newStopped = terminatedTasks.stream()
-                .filter(it -> !failureMonitor.hasFailed(it))
-                .map(TaskInfo::getName)
-                .collect(Collectors.toList());
+        List<TaskInfo> stopped = repairStatusRef.get().getStopped();
+        List<TaskInfo> failed = repairStatusRef.get().getFailed();
 
-        Optional<OfferRequirement> offerReq = Optional.empty();
-        boolean replacingFailed = false;
+        List<OfferRequirement> candidateRepairs = offerReqProvider.getTransientRecoveryOfferRequirements(stopped);
+        candidateRepairs.addAll(offerReqProvider.getPermanentRecoveryOfferRequirements(failed));
 
-        if (terminatedTasks.size() > 0) {
-            TaskInfo terminatedTask = terminatedTasks.get(new Random().nextInt(terminatedTasks.size()));
-            log.info("Found stopped task");
-            if (failureMonitor.hasFailed(terminatedTask)) {
-                log.info("Marking stopped task as failed");
-                failureListener.taskFailed(terminatedTask.getTaskId());
-            } else {
-                log.info("Relaunching stopped task in place");
-                offerReq = Optional.of(offerReqProvider.getReplacementOfferRequirement(terminatedTask));
-            }
-        } else {
-            log.info("Launching a failed task");
-            replacingFailed = true;
-            offerReq = offerReqProvider.maybeGetNewOfferRequirement(targetConfigName, block);
+        Optional<OfferRequirement> offerReq = Optional.empty();
+        if (candidateRepairs.size() > 0) {
+            offerReq = Optional.of(candidateRepairs.get(new Random().nextInt(candidateRepairs.size())));
```

Comment:
Will do.  The idea is that we're giving equal priority to both transiently and permanently failed tasks such that a flapping transient failure at the very least doesn't block recovery of permanently failed tasks.